### PR TITLE
Add payload format to LogEntryDescriptor.

### PR DESCRIPTION
**Pull request by douglas-reid**
At the moment, there is no way to distinguish the formats of various payloads sent via the payload attribute to mixer to enable better processing of the payloads.

This PR establishes support for structured payloads (serialized JSON). This will enable structured loggers to better include the payload in their generated logs.

### CODE REVIEW

File: mixer/v1/config/descriptor/log_entry_descriptor.proto [hunk 1]
```diff
@@ -33,4 +33,18 @@ message LogEntryDescriptor {
...
+
+    // PayloadFormat details the currently supported logging payload formats.
+    // TEXT is the default payload format.
+    enum PayloadFormat {
+
+        // Indicates a payload format of raw text.
+        TEXT = 0;
+
+        // Indicates a structured payload (JSON object).
+        STRUCTURED = 1;
```
Comment (mandarjog) [inline mixer/v1/config/descriptor/log_entry_descriptor.proto]: Does structured imply specifically json? or the fact that it has some structure ?

Comment (douglas-reid) [inline mixer/v1/config/descriptor/log_entry_descriptor.proto]: It implies JSON Object, serialized as string.  Do you prefer to call it JSON or something to be clearer?

Comment (wora) [inline mixer/v1/config/descriptor/log_entry_descriptor.proto]: It is likely we will add JSON to attribute type system, so we may not need this. Either payload format will be stored as string, so the format doesn't really matter much. We can simply document the type in the description.

If you use a log processing pipeline that doesn't handle JSON, you have to deal with it as text anyway.

If we do support JSON, this becomes a hard spec. It means the logging system must reject payload that is not valid JSON. Do you really want the logging system to valid the input? I doubt it is useful and it certainly adds high cost (memory and cpu).

Comment (douglas-reid) [inline mixer/v1/config/descriptor/log_entry_descriptor.proto]: Somewhere, we need something that documents the difference between raw text and JSON for the format of the string, especially if it gets re-serialized by the backend system. Documentation alone is not sufficient. In the prototype, the entries we received were Structs (or Protos or text). If we add JSON to the attribute types, we can remove this bit. But, in the meantime, I need something that can tell me what I am receiving so I can handle it appropriately.

Comment (wora) [inline mixer/v1/config/descriptor/log_entry_descriptor.proto]: There is subtle difference here. The log entry descriptor is for defining the log, such as I define an HTTP request log that uses JSON payload. If you write the mixer code, nobody can promise the caller of mixer will send you the valid json as oppose some garbage text.

However it does affect how you store the data, see https://github.com/googleapis/googleapis/blob/master/google/logging/v2/log_entry.proto#L73

I need to know what is your precise purpose so we can spec it out clearly. The platform spec must have very explicit and strict semantics.

Comment (douglas-reid) [inline mixer/v1/config/descriptor/log_entry_descriptor.proto]: This serves the same purpose the `json_payload` in the SCAPI you've linked.  The mixer can validate the JSON content, but it has to know that it _should_ be JSON.

My use case, at the moment, is simple. I have a basic logger that writes structured logs to stdout/stderr. I want to re-serialize the payload i receive as JSON appropriately.

**Comment by douglas-reid** [comment]
Renamed to JSON. PTAL.

**Comment by wora** [comment]
LGTM

**Pull request merged**